tools/gen_projects_json.py

```diff
@@ -1,300 +1,367 @@
-# get a token from here: https://github.com/settings/tokens
-
+import argparse
+import base64
+import datetime
+import json
 import os
 import re
 import sys
-import json
 import time
-import base64
 import urllib.request
-import datetime
+from pathlib import Path
 from pprint import pprint
 
 import yaml
 from boltons.urlutils import URL
-from boltons.fileutils import atomic_save
-from boltons.timeutils import isoparse
 
-TOOLS_PATH = os.path.dirname(os.path.abspath(__file__))
-PROJ_PATH = os.path.dirname(TOOLS_PATH)
-VTAG_RE = re.compile(r'''
+PROJECT_ROOT_PATH = Path(__file__).parent.parent
+VTAG_RE = re.compile(
+    r"""
     ^
     [^0-9]*
     (?P<major>\d+)
     \.
     [0-9a-zA-Z_.]+
-    ''', re.VERBOSE)
+    """,
+    re.VERBOSE,
+)
 
 # Tags matching these patterns will be completely skipped
 SKIP_PATTERNS = [
-    r'^ciflow/',  # pytorch has loads of this noise
-    r'^ci/',      # pytorch has loads of this noise
-    r'^nightly',  # FreeCol
+    r"^ciflow/",  # pytorch has loads of this noise
+    r"^ci/",  # pytorch has loads of this noise
+    r"^nightly",  # FreeCol
 ]
 
 # Version numbers after these patterns should be extracted
 STRIP_PATTERNS = [
-    r'^mc[0-9.]+-',  # Sodium tags include minecraft version numbers
+    r"^mc[0-9.]+-",  # Sodium tags include minecraft version numbers
 ]
 
-def strip_prefix(tag_name):
+
+def strip_prefix(tag_name: str) -> str:
     """Strip any non-numeric prefix from the tag name."""
-    # Handle cases where tag includes a path (e.g. refs/tags/v1.0.0)
-    _slash_prefix, _, tag_name = tag_name.rpartition('/')
-    
-    if '-' in tag_name:
-        prefix, _, version = tag_name.partition('-')
-        if re.search(r'^\d', version):
+    _, _, tag_name = tag_name.rpartition("/")
+
+    if "-" in tag_name:
+        _, _, version = tag_name.partition("-")
+        if re.search(r"^\d", version):
             return version
-    
-    match = re.search(r'\d', tag_name)
+
+    match = re.search(r"\d", tag_name)
     if match:
-        return tag_name[match.start():]
+        return tag_name[match.start() :]
     return tag_name
 
-def match_vtag(tag_name):
+
+def match_vtag(tag_name: str) -> re.Match | None:
     """Match version tags using a more general approach."""
     tag_name = strip_prefix(tag_name)
     return VTAG_RE.match(tag_name)
 
-def version_key(version):
+
+def version_key(version: str) -> tuple:
     """Extract and convert version numbers to tuple for comparison."""
     clean_version = strip_prefix(version)
-    # Split on any non-digit character and convert to integers
     try:
-        return tuple(int(x) for x in re.split(r'\D+', clean_version) if x and x.isdigit())
+        return tuple(
+            int(x) for x in re.split(r"\D+", clean_version) if x and x.isdigit()
+        )
     except (TypeError, ValueError):
         return tuple()
 
-def _get_gh_json(url):
+
+def _get_gh_json(url: str, args: argparse.Namespace) -> dict | list[dict]:
```

Comment:
As a pattern, I prefer to pass around a dataclass with the exact parameters vs an `argparse.Namespace` or dict